### PR TITLE
Add login, logout and verify powered by couchdb

### PR DESCRIPTION
**Pull request by leesper**
Hello Dirkjan, nice to meet you, recently I've been using couchdb-python in my project, and it offers me a great help, thank you! 
I find couchdb's "Cookie Authentication" and "Authentication Database" very useful, so I add some code in couchdb.client.Server to facilitate these things. Hope to make contribution to couchdb-python, thank you!

### CODE REVIEW

**Comment by djc** [comment]
Thanks for your contribution! Let me review your commits one by one...

**Comment by djc** [comment]
Let me know if you're okay with working on the stuff I mentioned! If not, I can clean stuff up a bit.

**Comment by djc** [comment]
Thanks for the improvements. I have one further request: please write some tests to cover these new methods.

**Comment by leesper** [comment]
ok, I will add some tests to cover these new methods, it may takes a few days since I have some refactoring work to do in my job, I will eventually write all tests you request, thanks a lot!

**Comment by leesper** [comment]
One last thing: I changed client.Server.veryfy_user() to support two authentication style: one is the 'token-based' before, the other is newly added, through username and password, so I rewrite the function a bit, thank you.

**Comment by leesper** [comment]
Mr. Dirkjan, I find something awful in my couchdb.Server.verify_user(self, token_or_name, password=None), and I correct it. First I don't really handle the exceptions raised when authenticated failed. It will raise ServerError if token-based authentication failed, or raise UnAuthorized when using username/password; second, username-password-based authentication should use POST and put the pair into body. All above have been corrected when I develop my production code.

**Comment by leesper** [comment]
I finally add a remove_user() in couchdb.client.Server, because I think if there's an add_user(), then there must be a remove_user(), and I add a little unit test for these user management thing, hope it meets your requirements, if there's something wrong, please tell me, I will keep correcting it until good, thank you very much.

File: couchdb/client.py [hunk 1]
```diff
@@ -275,16 +275,25 @@ def logout_user(self, token):
...
-            'Accept': 'application/json',
-            'Cookie': 'AuthSession=' + token,
-        }
+        if password is None:
+            header = {
+                'Accept': 'application/json',
+                'Cookie': 'AuthSession=' + token_or_name,
+            }
+        else:
+            from string import strip
```
Comment (djc) [inline couchdb/client.py]: Please don't use inline imports like this, just put them at the top of the file with the other imports.

File: couchdb/client.py [hunk 2]
```diff
@@ -275,16 +275,25 @@ def logout_user(self, token):
...
+            from base64 import encodestring
+            header = {
+                'Accept': 'application/json',
+                'Authorization': 'Basic ' + strip(encodestring(token_or_name + ':' + password)),
```
Comment (djc) [inline couchdb/client.py]: Nit: this line looks a bit long, maybe you can use a temporary variable for this.

File: couchdb/client.py [hunk 3]
```diff
@@ -255,12 +255,14 @@ def login_user(self, name, password):
             'name': name,
             'password': password,
         }
-        status, headers, _ = self.resource.post_json('_session', data)
-        if status != 200:
-            return status, None
-        cookie = headers.headers[0].split(';')[0]
-        pos = cookie.find('=')
-        return status, cookie[pos + 1:]
+        from http import Unauthorized
```
Comment (djc) [inline couchdb/client.py]: Imports to the top of the file, please.

File: couchdb/client.py [hunk 4]
```diff
@@ -284,19 +284,26 @@ def verify_user(self, token_or_name, password=None):
...
-            }
-        else:
-            from string import strip
-            from base64 import encodestring
-            header = {
-                'Accept': 'application/json',
-                'Authorization': 'Basic ' + strip(encodestring(token_or_name + ':' + password)),
-            }
-        status, _, _ = self.resource.get_json('_session', headers=header)
+        from http import Unauthorized, ServerError
```
Comment (djc) [inline couchdb/client.py]: Imports to the top.

File: couchdb/client.py [hunk 5]
```diff
@@ -284,19 +284,26 @@ def verify_user(self, token_or_name, password=None):
...
+            if password is None:
+                header = {
+                    'Accept': 'application/json',
+                    'Cookie': 'AuthSession=' + token_or_name,
+                }
+                status, _, _ = self.resource.get_json('_session', headers=header)
+            else:
+                header = {
+                    'Accept': 'application/json',
+                    'Authorization': 'Basic ' + strip(encodestring(token_or_name + ':' + password)),
```
Comment (djc) [inline couchdb/client.py]: If you're duplicating this logic in a few methods, perhaps you should extract some helper methods to create these header dicts?

File: couchdb/client.py [hunk 6]
```diff
@@ -297,9 +295,13 @@ def verify_user(self, token_or_name, password=None):
...
                     'Accept': 'application/json',
-                    'Authorization': 'Basic ' + strip(encodestring(token_or_name + ':' + password)),
+                    'Content-Type': 'application/json',
                 }
-                status, _, _ = self.resource.post_json('_session', headers=header)
+                body = {
+                    'name': token_or_name,
+                    'password': password,
+                }
+                status, _, _ = self.resource.post_json('_session', headers=header, body=body)
```
Comment (djc) [inline couchdb/client.py]: I don't think the `headers=` keyword argument needs to be explicitly named everywhere.

File: couchdb/client.py [hunk 7]
```diff
@@ -315,9 +314,7 @@ def verify_user(self, token_or_name, password=None):
                     'password': password,
                 }
                 status, _, _ = self.resource.post_json('_session', headers=header, body=body)
-        except Unauthorized:
-            return False
-        except ServerError:
+        except:
```
Comment (djc) [inline couchdb/client.py]: Please don't use bare `except:` clauses, and at least specify `Exception` here.

File: couchdb/client.py [hunk 8]
```diff
@@ -268,13 +268,12 @@ def login_user(self, name, password):
...
             'password': password,
         }
-        from http import Unauthorized
         try:
             status, headers, _ = self.resource.post_json('_session', data)
             cookie = headers.headers[0].split(';')[0]
             pos = cookie.find('=')
             return status, cookie[pos + 1:]
-        except Unauthorized:
+        except:
```
Comment (djc) [inline couchdb/client.py]: No bare `except`, please.

File: couchdb/client.py (hunk 6 above)
Comment (leesper) [inline couchdb/client.py]: ok, I will do some change

File: couchdb/tests/client.py [hunk 9]
```diff
@@ -145,9 +145,8 @@ def test_user_management(self):
         try:
             server.add_user('foo', 'secret', roles=['hero'])
             status, token = server.login_user('foo', 'secret')
-            self.assertEqual(status, 200)
-            self.assertTrue(server.verify_user(token))
-            self.assertTrue(server.logout_user(token))
+            server.verify_user(token)
```
Comment (djc) [inline couchdb/tests/client.py]: Why did you remove these? They seemed useful to me. The called methods can still return `False` on failure, right?

**Comment by leesper** [comment]
I'm very sorry, I still cannot pass the Travis ci test, I'm very sorry to commit quite a lot times  : (

**Comment by leesper** [comment]
I'll find another time figuring out how to pass the test!

**Comment by djc** [comment]
No worries, commits are cheap!

**Comment by leesper** [comment]
it seems that only py27 passed the tests, but py3x not, I still don't know how to deal with this, would you  please offer some tips? Thank you : )

**Comment by leesper** [comment]
I finally figure it out, in python3, there's no 'headers' field in headers returned by post_json(), instead it has a field called _header, that's why the integration tests failed, I correct it : )

File: couchdb/client.py [hunk 10]
```diff
@@ -270,10 +271,14 @@ def login_user(self, name, password):
...
-            cookie = headers.headers[0].split(';')[0]
+            if sys.version_info > (3, ):
+                cookie = headers._headers[0][1]
+            else:
+                cookie = headers.headers[0].split(';')[0]
             pos = cookie.find('=')
             return status, cookie[pos + 1:]
-        except Exception:
+        except Exception as e:
+            print(e)
```
Comment (iblis17) [inline couchdb/client.py]: Oops!

**Comment by djc** [comment]
I don't think printing the exception is the right solution, we're library and as such shouldn't send stuff to `stdout` or `stderr`. Maybe it would be better to just let the Exception bubble up to the caller.

**Comment by leesper** [comment]
And I think this would be appropriate: handling 'http Unauthorized' exception instead of just Exception, because the latter will swallow any other exception and the library user could hardly find what was wrong, just like me, spend quite a few time to figure out that there's difference between py2x and py3x

**Comment by djc** [comment]
Thanks for doing all this! I squashed your commits into a series that made a bit more sense to me, and then simplified it a bit more. This means the methods no longer catch exceptions, the names have changed a bit, and it can now only verify tokens. Hope that's okay to cover your use cases.

**Comment by leesper** [comment]
Mr. Dirkjan, I am wondering why it can only verify tokens? I am very curious about the reason. And of course, I think that is ok, although in my project I really use verify_user() to verify either username/password or token, well, I learned a lot in this issue, thank you very much. : )

**Comment by djc** [comment]
To me, it seemed like verifying the username/password combo duplicates functionality that's already implemented in the `login`/`logout` methods. Do you think this is not the case? Why do you need to verify username/password separately from being able to login?

**Comment by leesper** [comment]
It seems that everytime I call login() with a name and password, this function will always return me a new token, no matter whether I am already login or not, so sometimes when I take some action, I need to verify first, instead of getting new token, that's why I add it

**Comment by leesper** [comment]
In my product code, if a user wants to modify his/her password, then I need to verify whether the old one he/she offered is valid

**Comment by djc** [comment]
Sure, but if you have no other need of the token, you can just call logout with it again, right?

**Comment by leesper** [comment]
Yeah, never mind, I will adjust it when using in my way, thank you very much for your help and your patience. : )

**Pull request closed without merging**